netstack -> gvisor

reflog opened the pull request:
This PR restores the work done previously: https://github.com/getlantern/ipproxy/pull/11 in hopes that the new gvisor will not have issues with 32bit arch.

@hwh33 I'd appreciate a review if you can spare some cycles.

hwh33 commented on udp.go:
```diff
@@ -27,18 +28,20 @@ func (p *proxy) onUDP(pkt ipPacket) {
+	inboundPkt := stack.NewPacketBuffer(stack.PacketBufferOptions{
+		Payload: buffer.MakeWithData(pkt.raw),
 	})
+	defer inboundPkt.DecRef()
```
How do you know the packet is processed when this function exits?  Couldn't `InjectInbound` buffer it, say in a channel?

reflog replied:
@hwh33 just following the pattern in gvisor's code, like here: https://github.com/google/gvisor/blob/149350e5c428ae9b4ca8f2e2b7960071708113dd/pkg/tcpip/transport/tcp/testing/context/context.go#L502

hwh33 replied:
Okay, seems right then!

hwh33 commented on README.md:
```diff
@@ -6,6 +6,12 @@ A library which provides a proxy for IP traffic.
 
 This library uses Go modules. When running commands like `go test` in this repository, make sure the GO111MODULE environment variable is set to 'on'. See the [go command documentation](https://golang.org/cmd/go/#hdr-Preliminary_module_support) for more details. If you are running Go 1.13 or later, this should not be necessary as the Go tool will support modules by default.
 
+> In order to run ipproxy, the following tools need to be install: `lsof` and `ifconfig`.
+>
+> On Ubuntu: `apt install net-tools lsof`
+>
+> On Arch: `yay -S extra/lsof core/net-tools`
+
```
I think this is just because of `ipproxy.TUNDevice`, right?  That function is [only used in the demo](https://github.com/search?q=org%3Agetlantern+ipproxy.TUNDevice&type=code).  IMO, we should just make that a helper function in the demo code.  Then we won't have this requirement anymore and we can simplify the `ipproxy` API a bit.

reflog replied:
@hwh33 that code is also used by the tests. Not sure we want to remove it...

hwh33 replied:
Ah, thanks, I missed that!

hwh33 commented on base.go:
```diff
@@ -253,12 +260,16 @@ func (o *origin) copyToDownstream() {
-				return
-			case o.p.toDownstream <- pktInfo:
+		default:
+			if ptr := o.channelEndpoint.ReadContext(o.context); ptr != nil {
```
It looks like this is the only use of `baseConn.context` and the only need for `baseConn.cancelContext`.  [Here is an alternate approach](https://github.com/getlantern/ipproxy/pull/17/commits/04dc23b6c0e48406e71f01383c0d2057523f3178) which creates the context at the point of usage.  I thought it'd be easier to explain via code than to try to describe the idea in text!

hwh33 commented on ipproxy.go:
```diff
@@ -269,27 +280,29 @@ func (p *proxy) copyFromUpstream() {
-					return
-				case p.toDownstream <- pktInfo:
+			default:
+				if ptr := channelEndpoint.ReadContext(p.context); ptr != nil {
```
Similar to in the `baseConn` type, this is the only use of the `proxy` type's context.  [Here is an alternate approach for this context](https://github.com/getlantern/ipproxy/pull/17/commits/003fd07ee708e6d207f59264b7de5ff9a7912d34), again creating the context at the point of usage.

I also broke this goroutine out into a separate function.  This matches `baseConn`'s `copyToDownstream`, which serves a similar purpose.  It's also a bit odd IMO to call something which purports to just create a stack and find that it actually launched a goroutine which will copy packets back and forth.  And I realize that all of that predates this PR!

hwh33 commented on ipproxy.go:
```diff
@@ -269,27 +280,29 @@ func (p *proxy) copyFromUpstream() {
+					select {
+					case p.toDownstream <- ptr.Clone():
+						continue
+					default:
+						return
```
I fixed this [here](https://github.com/getlantern/ipproxy/pull/17/commits/003fd07ee708e6d207f59264b7de5ff9a7912d34#diff-9d4f61220fa14176fd93b67521aa8a039f15ac68514c97acfa011b3ecabbda49R284-R289), as part of [this change](https://github.com/getlantern/ipproxy/pull/16#discussion_r1432005989) simplifying our use of the serve context.

reflog commented:
@hwh33 amazing. thank you!

reflog commented on base.go:
```diff
@@ -96,25 +104,26 @@ func (conn *baseConn) getUpstream(timeout time.Duration) io.ReadWriteCloser {
 
 		if readErr != nil {
-			if readErr == tcpip.ErrWouldBlock {
+			if _, ok := readErr.(*tcpip.ErrWouldBlock); ok {
```
@hwh33 can't do that since readErr is not a real Go error, it's netstack one without Error() method

hwh33 replied:
Oh wow!  welp, never mind then 🙂

reflog commented:
hey @hwh33 thanks again, I addressed most of your comments and replied to those that I didn't change.
also - sorry for accusing you of breaking the build during last meeting, apparently, I had a broken tun0 on my system and it was screwing with the test :) mea culpa!

hwh33 reviewed (approved):
My only remaining question is [why we no longer need promiscuous mode in creating the TCP origin](https://github.com/getlantern/ipproxy/pull/16/files/45d01d38834834cf5a54607d6d00d41dbcdded34#r1431971021), but if it works without it, maybe it was never necessary.

Otherwise, looks good to me!

reflog commented:
@hwh33 to be honest - I don't know if SetPromiscous is required. It "seems" to work without it, but to avoid any weirdness - I've put it back in the code.

reflog commented:
i'll test flashlight with this version before merging and then push merge

reflog commented:
@hwh33 I've updated ipproxy in our lantern-client and seeing this:
```
E/GoLog   (29977): Jan 05 11:23:31.354 - 1m19s DEBUG ipproxy: ipproxy.go:207 Error on inbound packet, ignoring: Unsupported ip protocol version: 6
E/GoLog   (29977): Jan 05 11:23:31.355 - 1m19s DEBUG ipproxy: ipproxy.go:207 Error on inbound packet, ignoring: Unsupported ip protocol version: 6
E/GoLog   (29977): Jan 05 11:23:31.366 - 1m19s DEBUG ipproxy: ipproxy.go:207 Error on inbound packet, ignoring: Unsupported ip protocol version: 6
E/GoLog   (29977): Jan 05 11:23:31.590 - 1m19s DEBUG ipproxy: ipproxy.go:207 Error on inbound packet, ignoring: Unsupported ip protocol version: 6
E/GoLog   (29977): Jan 05 11:23:32.038 - 1m20s DEBUG ipproxy: ipproxy.go:207 Error on inbound packet, ignoring: Unsupported ip protocol version: 6
E/GoLog   (29977): Jan 05 11:23:35.240 - 1m23s DEBUG ipproxy: ipproxy.go:207 Error on inbound packet, ignoring: Unsupported ip protocol version: 6
I/GoLog   (29977): Jan 05 11:23:41.441 - 1m29s ERROR ipproxy: closeable.go:27 Use of idled network connection
I/GoLog   (29977): Jan 05 11:23:41.498 - 1m29s ERROR ipproxy: closeable.go:27 Use of idled network connection
I/GoLog   (29977): Jan 05 11:23:42.350 - 1m30s ERROR ipproxy: closeable.go:27 Use of idled network connection
E/GoLog   (29977): Jan 05 11:23:42.919 - 1m31s DEBUG ipproxy: ipproxy.go:207 Error on inbound packet, ignoring: Unsupported ip protocol version: 6
E/GoLog   (29977): Jan 05 11:23:46.354 - 1m34s DEBUG ipproxy: stats.go:36 TCP Origins: 12   TCP Clients: 3    UDP Conns: 3
E/GoLog   (29977): Jan 05 11:23:46.354 - 1m34s DEBUG ipproxy: stats.go:18 Accepted Packets: 92    Rejected Packets: 7
I/GoLog   (29977): Jan 05 11:23:56.725 - 1m44s ERROR ipproxy: closeable.go:27 Use of idled network connection
E/GoLog   (29977): Jan 05 11:23:57.767 - 1m45s DEBUG ipproxy: ipproxy.go:207 Error on inbound packet, ignoring: Unsupported ip protocol version: 6
I/GoLog   (29977): Jan 05 11:23:58.489 - 1m46s ERROR ipproxy: base.go:173 Unexpected error reading from upstream: Use of closed network connection [error=Unexpected error reading from upstream: %v error_location=github.com/getlantern/ipproxy.(*baseConn).copyFromUpstream (base.go:173) error_text=Unexpected error reading from upstream: Use of closed network connection error_type=errors.Error]
I/GoLog   (29977): ERROR ipproxy: base.go:173   at github.com/getlantern/ipproxy.(*baseConn).copyFromUpstream (base.go:173)
I/GoLog   (29977): ERROR ipproxy: base.go:173   at runtime.goexit (asm_arm64.s:1197)
I/GoLog   (29977): ERROR ipproxy: base.go:173 Caused by: Use of closed network connection
I/GoLog   (29977): Jan 05 11:23:58.490 - 1m46s ERROR ipproxy: closeable.go:27 Use of idled network connection
I/GoLog   (29977): Jan 05 11:24:00.067 - 1m48s ERROR ipproxy: closeable.go:27 Use of idled network connection
I/GoLog   (29977): Jan 05 11:24:00.068 - 1m48s ERROR ipproxy: closeable.go:27 Use of idled network connection
I/GoLog   (29977): Jan 05 11:24:00.069 - 1m48s ERROR ipproxy: closeable.go:27 Use of idled network connection
I/GoLog   (29977): Jan 05 11:24:00.072 - 1m48s ERROR ipproxy: closeable.go:27 Use of idled network connection
E/GoLog   (29977): Jan 05 11:24:01.353 - 1m49s DEBUG ipproxy: stats.go:36 TCP Origins: 16   TCP Clients: 5    UDP Conns: 15
E/GoLog   (29977): Jan 05 11:24:01.354 - 1m49s DEBUG ipproxy: stats.go:18 Accepted Packets: 164    Rejected Packets: 8
```


I thought that "Use of idled network connection" was hinting at an issue, but seems it was a red herring (below see the log of an older, working ipproxy).
At the moment I'm drawing a blank. Any ideas what could that be?

```
E/GoLog   (  375): Jan 05 11:38:56.004 - 0m48s DEBUG ipproxy: ipproxy.go:203 Error on inbound packet, ignoring: Unsupported ip protocol version: 6
E/GoLog   (  375): Jan 05 11:38:56.004 - 0m48s DEBUG ipproxy: ipproxy.go:203 Error on inbound packet, ignoring: Unsupported ip protocol version: 6
E/GoLog   (  375): Jan 05 11:38:56.006 - 0m48s DEBUG ipproxy: ipproxy.go:203 Error on inbound packet, ignoring: Unsupported ip protocol version: 6
I/GoLog   (  375): Jan 05 11:38:56.122 - 0m48s ERROR ipproxy: tcp.go:87 Unexpected error dialing upstream to 1.1.1.1:853: blackholing DNS over TLS traffic to 1.1.1.1:853 [error=Unexpected error dialing upstream to %v: %v error_location=github.com/getlantern/ipproxy.(*tcpOrigin).onAccept (tcp.go:87) error_text=Unexpected error dialing upstream to 1.1.1.1:853: blackholing DNS over TLS traffic to 1.1.1.1:853 error_type=errors.Error]
I/GoLog   (  375): ERROR ipproxy: tcp.go:87   at github.com/getlantern/ipproxy.(*tcpOrigin).onAccept (tcp.go:87)
I/GoLog   (  375): ERROR ipproxy: tcp.go:87   at runtime.goexit (asm_arm64.s:1165)
I/GoLog   (  375): ERROR ipproxy: tcp.go:87 Caused by: blackholing DNS over TLS traffic to 1.1.1.1:853
I/GoLog   (  375): ERROR ipproxy: tcp.go:87   at github.com/getlantern/android-lantern/internalsdk.Tun2Socks.func1 (tun.go:57)
I/GoLog   (  375): ERROR ipproxy: tcp.go:87   at github.com/getlantern/ipproxy.(*tcpOrigin).onAccept (tcp.go:85)
I/GoLog   (  375): ERROR ipproxy: tcp.go:87   at runtime.goexit (asm_arm64.s:1165)
E/GoLog   (  375): Jan 05 11:38:56.262 - 0m48s DEBUG ipproxy: ipproxy.go:203 Error on inbound packet, ignoring: Unsupported ip protocol version: 6
E/GoLog   (  375): Jan 05 11:38:56.390 - 0m49s DEBUG ipproxy: ipproxy.go:203 Error on inbound packet, ignoring: Unsupported ip protocol version: 6
E/GoLog   (  375): Jan 05 11:38:59.911 - 0m52s DEBUG ipproxy: ipproxy.go:203 Error on inbound packet, ignoring: Unsupported ip protocol version: 6
I/GoLog   (  375): Jan 05 11:39:06.092 - 0m58s ERROR ipproxy: closeable.go:27 Use of idled network connection
I/GoLog   (  375): Jan 05 11:39:06.126 - 0m58s ERROR ipproxy: closeable.go:27 Use of idled network connection
I/GoLog   (  375): Jan 05 11:39:06.147 - 0m58s ERROR ipproxy: closeable.go:27 Use of idled network connection
I/GoLog   (  375): Jan 05 11:39:06.294 - 0m58s ERROR ipproxy: closeable.go:27 Use of idled network connection
E/GoLog   (  375): Jan 05 11:39:07.590 - 1m0s DEBUG ipproxy: ipproxy.go:203 Error on inbound packet, ignoring: Unsupported ip protocol version: 6
I/GoLog   (  375): Jan 05 11:39:10.834 - 1m3s ERROR ipproxy: udp.go:53 Unable to dial upstream 240.0.0.10:443: blackholing QUIC UDP traffic to 240.0.0.10:443 [error=Unable to dial upstream %v: %v error_location=github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:53) error_text=Unable to dial upstream 240.0.0.10:443: blackholing QUIC UDP traffic to 240.0.0.10:443 error_type=errors.Error]
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:53)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at runtime.goexit (asm_arm64.s:1165)
I/GoLog   (  375): ERROR ipproxy: udp.go:53 Caused by: blackholing QUIC UDP traffic to 240.0.0.10:443
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/android-lantern/internalsdk.Tun2Socks.func2 (tun.go:73)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:49)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at runtime.goexit (asm_arm64.s:1165)
E/GoLog   (  375): Jan 05 11:39:11.004 - 1m3s DEBUG ipproxy: stats.go:28 TCP Origins: 8   TCP Clients: 9    UDP Conns: 9
E/GoLog   (  375): Jan 05 11:39:11.005 - 1m3s DEBUG ipproxy: stats.go:18 Accepted Packets: 170    Rejected Packets: 7
I/GoLog   (  375): Jan 05 11:39:11.627 - 1m4s ERROR ipproxy: udp.go:53 Unable to dial upstream 240.0.0.11:443: blackholing QUIC UDP traffic to 240.0.0.11:443 [error=Unable to dial upstream %v: %v error_location=github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:53) error_text=Unable to dial upstream 240.0.0.11:443: blackholing QUIC UDP traffic to 240.0.0.11:443 error_type=errors.Error]
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:53)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at runtime.goexit (asm_arm64.s:1165)
I/GoLog   (  375): ERROR ipproxy: udp.go:53 Caused by: blackholing QUIC UDP traffic to 240.0.0.11:443
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/android-lantern/internalsdk.Tun2Socks.func2 (tun.go:73)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:49)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at runtime.goexit (asm_arm64.s:1165)
I/GoLog   (  375): Jan 05 11:39:14.491 - 1m7s ERROR ipproxy: closeable.go:27 Use of idled network connection
I/GoLog   (  375): Jan 05 11:39:14.491 - 1m7s ERROR ipproxy: closeable.go:27 Use of idled network connection
I/GoLog   (  375): Jan 05 11:39:15.099 - 1m7s ERROR ipproxy: udp.go:53 Unable to dial upstream 240.0.0.5:443: blackholing QUIC UDP traffic to 240.0.0.5:443 [error=Unable to dial upstream %v: %v error_location=github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:53) error_text=Unable to dial upstream 240.0.0.5:443: blackholing QUIC UDP traffic to 240.0.0.5:443 error_type=errors.Error]
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:53)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at runtime.goexit (asm_arm64.s:1165)
I/GoLog   (  375): ERROR ipproxy: udp.go:53 Caused by: blackholing QUIC UDP traffic to 240.0.0.5:443
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/android-lantern/internalsdk.Tun2Socks.func2 (tun.go:73)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:49)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at runtime.goexit (asm_arm64.s:1165)
I/GoLog   (  375): Jan 05 11:39:15.290 - 1m7s ERROR ipproxy: udp.go:53 Unable to dial upstream 104.16.86.20:443: blackholing QUIC UDP traffic to 104.16.86.20:443 [error=Unable to dial upstream %v: %v error_location=github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:53) error_text=Unable to dial upstream 104.16.86.20:443: blackholing QUIC UDP traffic to 104.16.86.20:443 error_type=errors.Error]
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:53)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at runtime.goexit (asm_arm64.s:1165)
I/GoLog   (  375): ERROR ipproxy: udp.go:53 Caused by: blackholing QUIC UDP traffic to 104.16.86.20:443
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/android-lantern/internalsdk.Tun2Socks.func2 (tun.go:73)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:49)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at runtime.goexit (asm_arm64.s:1165)
I/GoLog   (  375): Jan 05 11:39:15.637 - 1m8s ERROR ipproxy: udp.go:53 Unable to dial upstream 240.0.0.11:443: blackholing QUIC UDP traffic to 240.0.0.11:443 [error=Unable to dial upstream %v: %v error_location=github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:53) error_text=Unable to dial upstream 240.0.0.11:443: blackholing QUIC UDP traffic to 240.0.0.11:443 error_type=errors.Error]
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:53)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at runtime.goexit (asm_arm64.s:1165)
I/GoLog   (  375): ERROR ipproxy: udp.go:53 Caused by: blackholing QUIC UDP traffic to 240.0.0.11:443
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/android-lantern/internalsdk.Tun2Socks.func2 (tun.go:73)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:49)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at runtime.goexit (asm_arm64.s:1165)
I/GoLog   (  375): Jan 05 11:39:15.724 - 1m8s ERROR ipproxy: udp.go:53 Unable to dial upstream 142.250.181.238:443: blackholing QUIC UDP traffic to 142.250.181.238:443 [error=Unable to dial upstream %v: %v error_location=github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:53) error_text=Unable to dial upstream 142.250.181.238:443: blackholing QUIC UDP traffic to 142.250.181.238:443 error_type=errors.Error]
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:53)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at runtime.goexit (asm_arm64.s:1165)
I/GoLog   (  375): ERROR ipproxy: udp.go:53 Caused by: blackholing QUIC UDP traffic to 142.250.181.238:443
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/android-lantern/internalsdk.Tun2Socks.func2 (tun.go:73)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:49)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at runtime.goexit (asm_arm64.s:1165)
I/GoLog   (  375): Jan 05 11:39:15.771 - 1m8s ERROR ipproxy: udp.go:53 Unable to dial upstream 104.16.86.20:443: blackholing QUIC UDP traffic to 104.16.86.20:443 [error=Unable to dial upstream %v: %v error_location=github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:53) error_text=Unable to dial upstream 104.16.86.20:443: blackholing QUIC UDP traffic to 104.16.86.20:443 error_type=errors.Error]
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:53)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at runtime.goexit (asm_arm64.s:1165)
I/GoLog   (  375): ERROR ipproxy: udp.go:53 Caused by: blackholing QUIC UDP traffic to 104.16.86.20:443
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/android-lantern/internalsdk.Tun2Socks.func2 (tun.go:73)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:49)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at runtime.goexit (asm_arm64.s:1165)
I/GoLog   (  375): Jan 05 11:39:16.420 - 1m9s ERROR ipproxy: udp.go:53 Unable to dial upstream 142.250.74.206:443: blackholing QUIC UDP traffic to 142.250.74.206:443 [error=Unable to dial upstream %v: %v error_location=github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:53) error_text=Unable to dial upstream 142.250.74.206:443: blackholing QUIC UDP traffic to 142.250.74.206:443 error_type=errors.Error]
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:53)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at runtime.goexit (asm_arm64.s:1165)
I/GoLog   (  375): ERROR ipproxy: udp.go:53 Caused by: blackholing QUIC UDP traffic to 142.250.74.206:443
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/android-lantern/internalsdk.Tun2Socks.func2 (tun.go:73)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:49)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at runtime.goexit (asm_arm64.s:1165)
I/GoLog   (  375): Jan 05 11:39:16.478 - 1m9s ERROR ipproxy: udp.go:53 Unable to dial upstream 142.250.184.251:443: blackholing QUIC UDP traffic to 142.250.184.251:443 [error=Unable to dial upstream %v: %v error_location=github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:53) error_text=Unable to dial upstream 142.250.184.251:443: blackholing QUIC UDP traffic to 142.250.184.251:443 error_type=errors.Error]
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:53)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at runtime.goexit (asm_arm64.s:1165)
I/GoLog   (  375): ERROR ipproxy: udp.go:53 Caused by: blackholing QUIC UDP traffic to 142.250.184.251:443
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/android-lantern/internalsdk.Tun2Socks.func2 (tun.go:73)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at github.com/getlantern/ipproxy.(*proxy).startUDPConn.func2 (udp.go:49)
I/GoLog   (  375): ERROR ipproxy: udp.go:53   at runtime.goexit (asm_arm64.s:1165)
```

hwh33 commented:
Hey sorry @reflog, just saw this.  I notice that in the older logs, we see messages about black-holing UDP traffic.  Is it possible that this black-holing is no longer happening?  I think it's necessary as Lantern proxies won't handle UDP traffic.

reflog commented:
@hwh33 I'll look into it, though I have no idea what blackholing UDP means in this context :)

hwh33 commented:
I'll be honest, I don't know a ton about it either 🙃 

I think here it just means returning an error indiscriminately for such traffic.  [Here is the source of these errors](https://github.com/getlantern/lantern-client/blob/771a1ac79c68a75025bfd145d2cba8c0c13100b3/internalsdk/tun.go#L54-L74).  If I recall correctly from conversations with Ox, this is necessary to force services on the device to use TCP.  This is particularly important for DNS.  But don't quote me on any of this!

atavism commented:
@reflog Does the demo code still work with your changes? I wonder if we could get that working on macOS to do some further testing/investigation what the issue might be on Android. I guess we need to create a tun device like [here](https://github.com/getlantern/ipproxy/blob/master/tun_linux.go#L14)?
